codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -209,7 +209,15 @@ public void operationComplete(ChannelFuture future) throws Exception {
                 if (failureCause == null) {
                     // Synchronously set the headersSent flag to ensure that we do not subsequently write
                     // other headers containing pseudo-header fields.
+                    //
+                    // This is best-afford as the operation is async it may still fail later on. We just update it here
+                    // directly as otherwise it may be challenging for the end-user to co-ordinate writes that depend
+                    // on the state.
                     stream.headersSent(isInformational);
+
+                    if (!future.isDone()) {
```

Comment:
Hmm, perhaps you can move this check into `notifyLifecycleManagerOnError()`?

Reply:
sure

Reply:
done